src/Autofac/Features/Decorators/DecoratorMiddleware.cs

```diff
@@ -0,0 +1,127 @@
+﻿// This software is part of the Autofac IoC container
+// Copyright © 2020 Autofac Contributors
+// https://autofac.org
+//
+// Permission is hereby granted, free of charge, to any person
+// obtaining a copy of this software and associated documentation
+// files (the "Software"), to deal in the Software without
+// restriction, including without limitation the rights to use,
+// copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+// OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+// OTHER DEALINGS IN THE SOFTWARE.
+
+using System;
+using System.Linq;
+using Autofac.Core;
+using Autofac.Core.Resolving.Pipeline;
+
+namespace Autofac.Features.Decorators
+{
+    /// <summary>
+    /// Service middleware that decorates activated instances of a service.
+    /// </summary>
+    internal class DecoratorMiddleware : IResolveMiddleware
+    {
+        private readonly DecoratorService _decoratorService;
+        private readonly IComponentRegistration _decoratorRegistration;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DecoratorMiddleware"/> class.
+        /// </summary>
+        /// <param name="decoratorService">The decorator service.</param>
+        /// <param name="decoratorRegistration">The decorator registration.</param>
+        public DecoratorMiddleware(DecoratorService decoratorService, IComponentRegistration decoratorRegistration)
+        {
+            _decoratorService = decoratorService;
```

Comment:
Things will break if these are null, though less of a risk on an internal class.

Reply:
(I think there are a couple of new classes where that's the case. I'm not super concerned if you're not, given they're internal. The public classes probably do need null checks.)

Reply:
I think I've got them all.